chain/p2p_test.go

```diff
@@ -0,0 +1,392 @@
+package chain
+
+import (
+	"bytes"
+	"errors"
+	"net"
+	"strings"
+	"testing"
+	"time"
+
+	"go.sia.tech/core/consensus"
+	"go.sia.tech/core/gateway"
+	"go.sia.tech/core/types"
+)
+
+func blockEqual(b1, b2 types.Block) bool {
+	var buf1 bytes.Buffer
+	{
+		e := types.NewEncoder(&buf1)
+		types.V2Block(b1).EncodeTo(e)
+		e.Flush()
+	}
+
+	var buf2 bytes.Buffer
+	{
+		e := types.NewEncoder(&buf2)
+		types.V2Block(b2).EncodeTo(e)
+		e.Flush()
+	}
+
+	return bytes.Equal(buf1.Bytes(), buf2.Bytes())
+}
+
+func testNetwork() (*consensus.Network, types.Block) {
+	// use a modified version of Zen
+	n, genesisBlock := TestnetZen()
+	n.InitialTarget = types.BlockID{0xFF}
+	n.HardforkDevAddr.Height = 1
+	n.HardforkTax.Height = 1
+	n.HardforkStorageProof.Height = 1
+	n.HardforkOak.Height = 1
+	n.HardforkASIC.Height = 1
+	n.HardforkFoundation.Height = 1
+	n.HardforkV2.AllowHeight = 5
+	n.HardforkV2.RequireHeight = 10
+	return n, genesisBlock
+}
+
+type subscriber struct {
+	update *ApplyUpdate
+}
+
+func (s *subscriber) ProcessChainApplyUpdate(cau *ApplyUpdate, _ bool) error {
+	s.update = cau
+	return nil
+}
+
+func (s *subscriber) ProcessChainRevertUpdate(cru *RevertUpdate) error {
+	return nil
+}
+
+type handler struct{ cm *Manager }
+
+func (h handler) PeersForShare() []string { return nil }
+func (h handler) Block(id types.BlockID) (types.Block, error) {
+	block, ok := h.cm.Block(id)
+	if !ok {
+		return types.Block{}, errors.New("no such block")
+	}
+	return block, nil
+}
+func (h handler) BlocksForHistory(history []types.BlockID, max uint64) ([]types.Block, uint64, error) {
+	fill := max
+	if max > uint64(len(history)) {
+		fill = uint64(len(history))
+	}
+	remaining := uint64(0)
+	if max > fill {
+		remaining = max - fill
+	}
+
+	var blocks []types.Block
+	for i := uint64(0); i < fill; i++ {
+		block, err := h.Block(history[i])
+		if err != nil {
+			return nil, 0, err
+		}
+		blocks = append(blocks, block)
+	}
+	return blocks, remaining, nil
+}
+func (h handler) RelayHeader(gateway.BlockHeader, *gateway.Peer)         {}
+func (h handler) RelayTransactionSet([]types.Transaction, *gateway.Peer) {}
+func (h handler) Transactions(types.ChainIndex, []types.Hash256) ([]types.Transaction, []types.V2Transaction, error) {
+	return nil, nil, nil
+}
+func (h handler) Checkpoint(types.ChainIndex) (types.Block, consensus.State, error) {
+	return types.Block{}, consensus.State{}, nil
+}
+func (h handler) RelayV2Header(gateway.V2BlockHeader, *gateway.Peer)         {}
+func (h handler) RelayV2BlockOutline(gateway.V2BlockOutline, *gateway.Peer)  {}
+func (h handler) RelayV2TransactionSet([]types.V2Transaction, *gateway.Peer) {}
+
+func TestP2P(t *testing.T) {
```

Comment:
so the bad news is that this test only adds 1% of coverage to `chain`. Not worth it imo. I know it was a pain to get it working, sorry :(

The good news is that this test *will* be nice to have if we end up moving more syncer code into `core`. So we should keep it around in the git history for that, at least.

I think the best place to focus on for now is file contract tests in `consensus` (v1 and v2). It's crucial that we test as many edge cases as possible in the new consensus code, so we should strive for 100% coverage.